### PR TITLE
Use the same naming rules for the names of logging dimensions

### PR DESCRIPTION
**Pull request by iliar-turdushev**
The following has been implemented:

* Use [OTel semantic conventions](https://github.com/open-telemetry/semantic-conventions/tree/b127e121b9b458e83a0fee8a32f7eb6721912951) for the names of logging dimensions where possible. Since [semantic conventions for logs](https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/general/logs.md) are **Experimental**, and are missed for the most of areas/objects we need to agree on the dimensions names, we decided to use [semantic conventions for traces](https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/general/trace.md). Semantic conventions for traces are also **Experimental**, but they are defined for a wide variety of areas/objects.
* If a semantic convention for a dimensions or a log message parameter is missing, then we use `PascalCase` for the name of the dimension.
* Use dot "." instead of underscore "_" as a separator for the names of dimension of `LogPropertiesAttribute` and `TagProviderAttribute`.
* Remove "env_" preffix from the names of enrichment dimensions.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4545)

### CODE REVIEW

**Comment by iliar-turdushev** [comment]
@ananthsr If we [rename](https://github.com/dotnet/extensions/blob/main/src/Libraries/Microsoft.Extensions.Http.Diagnostics/Latency/Internal/HttpClientLatencyLogEnricher.cs#L53C13-L54C1) enrichment property `latencyInfo` to `LatencyInfo` (the first letter is Capital), won't it affect latency measurement? Thank you.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs:19]:
```suggestion
        $"{{HttpClientLoggingTagNames.Method}} {{HttpClientLoggingTagNames.Host}}/{{HttpClientLoggingTagNames.Path}}";
```

**Comment by klauco** [comment]
> @ananthsr If we [rename](https://github.com/dotnet/extensions/blob/main/src/Libraries/Microsoft.Extensions.Http.Diagnostics/Latency/Internal/HttpClientLatencyLogEnricher.cs#L53C13-L54C1) enrichment property `latencyInfo` to `LatencyInfo` (the first letter is Capital), won't it affect latency measurement? Thank you.

let's ensure we align the naming here regardless of the impact.

**Review by Tratcher — APPROVED** [review]

File: src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Enrichment/ApplicationEnricherTags.cs
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Enrichment/ApplicationEnricherTags.cs:17]: Is this lowercased because it's used in metrics only?

File: src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Enrichment/ProcessEnricherTagNames.cs
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Enrichment/ProcessEnricherTagNames.cs:22]: Same question applies here

**Comment by xakep139** [comment]
Is the PR still draft?

**Comment by xakep139** [comment]
@iliar-turdushev any chance you checked logging categories we use?

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (iliar-turdushev) [inline src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:22]: New name https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/http/http-spans.md#http-server-semantic-conventions

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (iliar-turdushev) [inline src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:37]: New name [http.request.header.](https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/http/http-spans.md#common-attributes)

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (iliar-turdushev) [inline src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:42]: New name [http.request.header.](https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/http/http-spans.md#common-attributes)

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (xakep139) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:22]: You meant removal of `http`?

File: src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Enrichment/ApplicationEnricherTags.cs
Comment (xakep139) [reply src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Enrichment/ApplicationEnricherTags.cs:17]: What do you think about putting the link into the code (as a comment)?

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (xakep139) [reply src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:37]: We also need to normalize header names:
image

Comment (xakep139) [reply src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:37]: It's defined in https://github.com/dotnet/extensions/blob/61a28bead089f12396158b52bca00df48847a732/src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HeaderReader.cs#L40

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (xakep139) [reply src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:42]: See the comment above

Comment (xakep139) [reply src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:42]: Wait, the link you shared is related to spans. Is it applicable to logs?

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (iliar-turdushev) [inline src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:1]: Some dimensions (e.g. `Host` and `Method`) were not renamed, though they have corresponding names in [OTel's semantic conventions for HTTP spans](https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/http/http-spans.md#semantic-conventions-for-http-spans). The reason is that they are coming from ASP.NET Core 8's HttpLogging, therefore we cannot rename them.

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:37]: Yeah, at first I just wanted to agree if we are going to follow this conventions, and then make such changes.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:22]: Correct, `http`.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (iliar-turdushev) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:1]: We were using [Semantic convetions for HTTP Spans](https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/http/http-spans.md#semantic-conventions-for-http-spans) for choosing names for the dimensions.

`Duration`, `RequestBody` and `ResponseBody` doesn't have matching semantic conventions. There is a metric for duration of Http Client requests though. But it's a [metric name](https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/http/http-metrics.md#metric-httpclientrequestduration) rather than a dimension name, therefore we didn't apply it.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs
Comment (iliar-turdushev) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs:49]: New names of HttpClientLogging dimensions contain dots, e.g. `http.request.method`, but LoggerMessage currently don't support such parameters, because the name of a parameter must match the name of an argument of the logging method. Therefore, in order to create a log message with dimensions containing dots, we have to create the message manually.

The code below is a copy for code generated by `LoggerMessage`, I only made changes to the parameter names.

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/HeaderParsingFeature.cs
Comment (iliar-turdushev) [inline src/Libraries/Microsoft.AspNetCore.HeaderParsing/HeaderParsingFeature.cs:201]: I think we shouldn't apply a name [`http.request.header.<key>`](https://github.com/open-telemetry/semantic-conventions/blob/b127e121b9b458e83a0fee8a32f7eb6721912951/docs/attributes-registry/http.md) here, because we are not logging the value of the header here. The aim of this and other logger messages in this file is to log the name of the header, not its value, i.e. the data we should store in a log record should be `[HeaderName] = "name_of_the_header` rather than `["name_of_the_header"] = ???`.

What do you think?

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs]: Closing this discussion, since it is not relevant after the latest changes.

File: src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Enrichment/ApplicationEnricherTags.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Enrichment/ApplicationEnricherTags.cs:17]: That's a good idea. When we agree on the names, I'll put appropriate links into our code.

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:37]: Probably, we'll also need to apply the request header prefix and normalization in the RequestHeaderLogEnricher:
https://github.com/dotnet/extensions/blob/6fda2c8b8f8bacf74eaafff243a10ea20f849aaa/src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/RequestHeadersLogEnricher.cs#L68

**Comment by iliar-turdushev** [comment]
> @iliar-turdushev any chance you checked logging categories we use?

@joperezr You asked to check logging categories https://github.com/dotnet/extensions/issues/4451#issuecomment-1749231823.

I checked them, and they look good. What exactly I should pay attention to? Who can give guidance?

Thank you.

**Comment by iliar-turdushev** [comment]
@reyang @noahfalk Please, review this PR. Thank you.

File: src/Generators/Microsoft.Gen.Logging/Emission/Emitter.Method.cs
Comment (geeknoid) [inline src/Generators/Microsoft.Gen.Logging/Emission/Emitter.Method.cs:18]: I'm not thrilled with this NUllablePropertySeparator. PropertySeparator is used when we're producing a name string to be used in telemetry, so the string could in fact be anything we want and not break anything. However, NullablePropertySeparator is used to produce source code that references a particular member in the generated code. So that can't readily be changed as it would break the generated code.

Maybe just add some comments to deacribe what these are for...

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/HeaderParsingFeature.cs
Comment (geeknoid) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/HeaderParsingFeature.cs:201]: Agreed.

Why change the logging messages from headerName to HeaderName?

**Review by geeknoid — COMMENTED** [review]
Why do we need to change all the logging templates from {foo} to {Foo}?

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs
Comment (geeknoid) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs:49]: Let's file a feature request to introduce a [TagName] attribute that the code generator will take the final name from.

**Comment by iliar-turdushev** [comment]
> Why do we need to change all the logging templates from {foo} to {Foo}?

One of the goals of this PR is to be aligned with naming rules/conventions being used in the rest of dotnet, in case there is no a corresponding OTel semantic-convention. In dotnet a `PascalCase` is being used for the names of logging dimensions and parameters of logger messages. For example: https://github.com/dotnet/aspnetcore/blob/main/src/Middleware/HttpLogging/src/HttpLoggingExtensions.cs#L23C1-L24C93.

File: src/Generators/Microsoft.Gen.Logging/Emission/Emitter.Method.cs
Comment (iliar-turdushev) [reply src/Generators/Microsoft.Gen.Logging/Emission/Emitter.Method.cs]: I didn't know that the string constants were inlined intentionally, therefore when making changes I extracted them into constant fields following "best practices". I removed the constant fields and inlined the constant strings.

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/HeaderParsingFeature.cs
Comment (xakep139) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/HeaderParsingFeature.cs:201]: +1

File: src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs
Comment (xakep139) [reply src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Logging/HttpLoggingTagNames.cs:37]: If we agree on the change, it should be also applied to HttpClient logging of headers

**Comment by xakep139** [comment]
@iliar-turdushev the PR LGTM, the only missing bit is header names normalization (if we agree on the naming)

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs:49]: Create an issue https://github.com/dotnet/extensions/issues/4576.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (dariusclay) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:42]: even though it's not part of semantic conventions, why aren't we following the same naming rules? ie: `http.response.body`

**Review by geeknoid — APPROVED** [review]

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:42]: That's what we agreed on:
* if a dimension is a part of the semantic conventions, then we use it;
* otherwise - we use dotnet naming rules for logging dimensions, i.e. `PascalCase`.

**Comment by joperezr** [comment]
@xakep139 @iliar-turdushev Other than fixing code coverage, is this ready to go in?

**Comment by iliar-turdushev** [comment]
> @xakep139 @iliar-turdushev Other than fixing code coverage, is this ready to go in?

@joperezr We are waiting for review from @reyang and @noahfalk. Also, if we agree to follow OTel's HTTP Request/Response header name semantics, then we'll also need to implement header name normalization.

**Review by noahfalk — APPROVED** [review]
This seems fine to me, but for anything nuanced about the conventions I probably wouldn't catch it. If you want a higher level of scrutiny, someone like Liudmila is probably a better reviewer than I.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Microsoft.Extensions.Http.Diagnostics.csproj
Comment (iliar-turdushev) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Microsoft.Extensions.Http.Diagnostics.csproj:30]: Code coverage has dropped because we had to replace code generated logger messages with their actual implementations (i.e. see [here](https://github.com/dotnet/extensions/pull/4545/files#diff-01befa772fdd79453ebf661760fa2509b4cfc9f0cbe25c7644cc1680c8678ef1R129)) because of the following limitation of the `LoggerMessage` #4576. As soon as we implement #4576, we increase the code coverage again.

**Comment by iliar-turdushev** [comment]
> This seems fine to me, but for anything nuanced about the conventions I probably wouldn't catch it. If you want a higher level of scrutiny, someone like Liudmila is probably a better reviewer than I.

@lmolkova Please, review the PR. Thank you.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (lmolkova) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:22]: note that HTTP Host header is a combination of host and port. If you want to capture the host header as it, it's best to use `http.request.header.host` attribute instead.

Also, `server.address` is not always derived from the host header. It may be derived from `Forwareded` or `X-Forwarded-Host`.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (lmolkova) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:42]: It looks inconsistent and it'd be great to have the same naming conventions across different attributes. This way user at least won't need to think about casing when writing their queries against logs.

As a side-note, OTel is now incorporating a lot of attributes from ECS - https://www.elastic.co/guide/en/ecs/current/ecs-http.html. There is no promise they will go as is, but they have a lot of relevant attribute such as [http.request.body.bytes](https://www.elastic.co/guide/en/ecs/current/ecs-http.html#field-http-request-body-bytes) or [http.request.body.content](https://www.elastic.co/guide/en/ecs/current/ecs-http.html#field-http-request-body-content)

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs
Comment (lmolkova) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs:57]: would be nice to add [exception.type](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/exceptions/exceptions-logs.md)  and other `exception` attributes

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs
Comment (lmolkova) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs:84]: same here and in other places where exception is reported - `exception.*` attributes would be awesome

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/LoggerMessageStateExtensions.cs
Comment (lmolkova) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/LoggerMessageStateExtensions.cs:60]: same here - no need for `Replace('-', '_')` anymore

**Review by lmolkova — COMMENTED** [review]
Left a few minor comments on naming.

I'm a bit concerned about the inconsistent naming patterns. otel.attributes and PascalCaseAttributes in the same log record would look messy.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (lmolkova) [inline src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:42]: It looks inconsistent and it'd be great to have the same naming conventions across different attributes. This way user at least won't need to think about casing when writing their queries against logs.

As a side-note, OTel is now incorporating a lot of attributes from ECS - https://www.elastic.co/guide/en/ecs/current/ecs-http.html. There is no promise they will go as is, but they have a lot of relevant attribute such as [http.request.body.bytes](https://www.elastic.co/guide/en/ecs/current/ecs-http.html#field-http-request-body-bytes) or [http.request.body.content](https://www.elastic.co/guide/en/ecs/current/ecs-http.html#field-http-request-body-content)

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs:57]: This was implemented in another PR #4584.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/Log.cs:84]: This was implemented in another PR #4584.

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:22]: We are reading the host from a request's Host property which is the name of the host without port information:

https://github.com/dotnet/extensions/blob/d74d9f4915cf6a1bdd38f784fcbda2861312bd8e/src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/HttpRequestReader.cs#L93

So, as I understand, in our case `Host` is exactly `server.address` dimensions from the semantic-convetions.

**Comment by iliar-turdushev** [comment]
> Left a few minor comments on naming.
> 
> I'm a bit concerned about the inconsistent naming patterns. otel.attributes and PascalCaseAttributes in the same log record would look messy.

@lmolkova Thank you very much for reviewing the PR.

Yeah, we agree, but, unfortunately, there is no perfect solution here. In our opinion, the best we can do is to follow OTel's semantic-conventions where possible. The rest, which is not yet present in OTel or cannot be aligned with OTel due to implementation details in .NET, will be aligned with .NET's naming rules. This, at least, will allow us to leverage some benefits of following OTel's schema, for example, integration with tools built on top of it.

cc @klauco

**Review by lmolkova — APPROVED** [review]

File: src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs
Comment (iliar-turdushev) [reply src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/HttpClientLoggingTagNames.cs:42]: > As a side-note, OTel is now incorporating a lot of attributes from ECS - https://www.elastic.co/guide/en/ecs/current/ecs-http.html. There is no promise they will go as is, but they have a lot of relevant attribute such as [http.request.body.bytes](https://www.elastic.co/guide/en/ecs/current/ecs-http.html#field-http-request-body-bytes) or [http.request.body.content](https://www.elastic.co/guide/en/ecs/current/ecs-http.html#field-http-request-body-content)

Thank you for the information. For now we want to follow what OTel conventions provide, and don't want not to bring confusion by following both OTel and ECS, since if we bring some names from ECS then someone can reasonably ask why we are following ECS here, and not there.

**Pull request merged**
